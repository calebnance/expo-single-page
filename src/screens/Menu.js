import React from 'react';
import { Animated, ScrollView, StyleSheet, View } from 'react-native';
import PropTypes from 'prop-types';
import { colors, device } from '../constants';

// components
import Touch from '../components/Touch';

class Menu extends React.Component {
  constructor(props) {
    super(props);

    let positionStart = -device.width;
    if (props.direction === 'right') {
      positionStart = device.width;
    }

    this.state = {
      backgroundColor: new Animated.Value(0),
      position: new Animated.Value(positionStart),
      positionStart
    };

    this.handleClose = this.handleClose.bind(this);
  }

  componentDidMount() {
    const { backgroundColor, position } = this.state;

    Animated.timing(backgroundColor, {
      delay: 100,
      duration: 400,
      toValue: 10
    }).start();

    Animated.timing(position, {
      duration: 400,
      toValue: 0
    }).start();
  }

  handleClose() {
    const { onClose } = this.props;
    const { backgroundColor, position, positionStart } = this.state;

    Animated.timing(backgroundColor, {
      duration: 100,
      toValue: 0
    }).start();

    Animated.timing(position, {
      duration: 400,
      toValue: positionStart
    }).start(() => {
      onClose();
    });
  }

  renderLayout() {
    const { direction } = this.props;
    const { backgroundColor } = this.state;

    let render = '';
    const color = backgroundColor.interpolate({
      inputRange: [0, 10],
      outputRange: [colors.white0, colors.white20]
    });

    const mockItems = [];

    for (let i = 0; i < 20; i += 1) {
      mockItems.push(<View key={i} style={styles.mockBlock} />);
    }

    const scrollViewArea = (
      <ScrollView style={styles.scrollView}>
        {mockItems}
        <View style={styles.listSpacer} />
      </ScrollView>
    );

    const closeArea = (
      <Touch onPress={() => this.handleClose()} style={styles.closeArea} />
    );

    if (direction === 'left') {
      render = (
        <Animated.View style={[styles.panel, { backgroundColor: color }]}>
          {scrollViewArea}
          {closeArea}
        </Animated.View>
      );
    } else if (direction === 'right') {
      render = (
        <Animated.View style={[styles.panel, { backgroundColor: color }]}>
          {closeArea}
          {scrollViewArea}
        </Animated.View>
      );
    }

    return render;
  }

  render() {
    const { show } = this.props;
    const { position } = this.state;

    if (show === false) {
      return null;
    }

    return (
      <Animated.View style={[styles.container, { left: position }]}>
        {this.renderLayout()}
      </Animated.View>
    );
  }
}

// default props
Menu.defaultProps = {
  direction: 'left',
  show: false
};

// type checking
Menu.propTypes = {
  // required
  onClose: PropTypes.func.isRequired,

  // optional
  direction: PropTypes.string,
  show: PropTypes.bool
};

const styles = StyleSheet.create({
  container: {
    height: '100%',
    position: 'absolute',
    width: '100%',
    zIndex: 50
  },
  panel: {
    flex: 1,
    flexDirection: 'row'
  },
  scrollView: {
    backgroundColor: colors.brown,
    paddingTop: device.iPhoneX ? 44 : 0,
    width: '80%'
  },
  closeArea: {
    height: '100%',
    width: '20%'
  },
  mockBlock: {
    backgroundColor: colors.grey,
    height: 75,
    marginBottom: 20,
    width: '100%'
  },
  listSpacer: {
    height: device.iPhoneX ? 44 : 0
  }
});

export default Menu;
