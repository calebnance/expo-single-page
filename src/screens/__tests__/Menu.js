import React from 'react';
import renderer from 'react-test-renderer';

import Menu from '../Menu';

describe('<Menu />', () => {
  beforeEach(() => {
    jest.useFakeTimers();
  });

  it('is hidden', () => {
    const tree = renderer.create(<Menu onClose={jest.fn()} />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders correctly on left', () => {
    const tree = renderer.create(<Menu onClose={jest.fn()} show />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders correctly on right', () => {
    const tree = renderer
      .create(<Menu direction="right" onClose={jest.fn()} show />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
