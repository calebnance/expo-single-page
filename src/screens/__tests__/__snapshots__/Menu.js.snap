// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Menu /> is hidden 1`] = `null`;

exports[`<Menu /> renders correctly on left 1`] = `
<View
  style={
    Object {
      "height": "100%",
      "left": -750,
      "position": "absolute",
      "width": "100%",
      "zIndex": 50,
    }
  }
>
  <View
    style={
      Object {
        "backgroundColor": "rgba(255, 255, 255, 0)",
        "flex": 1,
        "flexDirection": "row",
      }
    }
  >
    <RCTScrollView
      style={
        Object {
          "backgroundColor": "#bcab79",
          "paddingTop": 0,
          "width": "80%",
        }
      }
    >
      <View>
        <View
          style={
            Object {
              "backgroundColor": "#d0ccd0",
              "height": 75,
              "marginBottom": 20,
              "width": "100%",
            }
          }
        />
        <View
          style={
            Object {
              "backgroundColor": "#d0ccd0",
              "height": 75,
              "marginBottom": 20,
              "width": "100%",
            }
          }
        />
        <View
          style={
            Object {
              "backgroundColor": "#d0ccd0",
              "height": 75,
              "marginBottom": 20,
              "width": "100%",
            }
          }
        />
        <View
          style={
            Object {
              "backgroundColor": "#d0ccd0",
              "height": 75,
              "marginBottom": 20,
              "width": "100%",
            }
          }
        />
        <View
          style={
            Object {
              "backgroundColor": "#d0ccd0",
              "height": 75,
              "marginBottom": 20,
              "width": "100%",
            }
          }
        />
        <View
          style={
            Object {
              "backgroundColor": "#d0ccd0",
              "height": 75,
              "marginBottom": 20,
              "width": "100%",
            }
          }
        />
        <View
          style={
            Object {
              "backgroundColor": "#d0ccd0",
              "height": 75,
              "marginBottom": 20,
              "width": "100%",
            }
          }
        />
        <View
          style={
            Object {
              "backgroundColor": "#d0ccd0",
              "height": 75,
              "marginBottom": 20,
              "width": "100%",
            }
          }
        />
        <View
          style={
            Object {
              "backgroundColor": "#d0ccd0",
              "height": 75,
              "marginBottom": 20,
              "width": "100%",
            }
          }
        />
        <View
          style={
            Object {
              "backgroundColor": "#d0ccd0",
              "height": 75,
              "marginBottom": 20,
              "width": "100%",
            }
          }
        />
        <View
          style={
            Object {
              "backgroundColor": "#d0ccd0",
              "height": 75,
              "marginBottom": 20,
              "width": "100%",
            }
          }
        />
        <View
          style={
            Object {
              "backgroundColor": "#d0ccd0",
              "height": 75,
              "marginBottom": 20,
              "width": "100%",
            }
          }
        />
        <View
          style={
            Object {
              "backgroundColor": "#d0ccd0",
              "height": 75,
              "marginBottom": 20,
              "width": "100%",
            }
          }
        />
        <View
          style={
            Object {
              "backgroundColor": "#d0ccd0",
              "height": 75,
              "marginBottom": 20,
              "width": "100%",
            }
          }
        />
        <View
          style={
            Object {
              "backgroundColor": "#d0ccd0",
              "height": 75,
              "marginBottom": 20,
              "width": "100%",
            }
          }
        />
        <View
          style={
            Object {
              "backgroundColor": "#d0ccd0",
              "height": 75,
              "marginBottom": 20,
              "width": "100%",
            }
          }
        />
        <View
          style={
            Object {
              "backgroundColor": "#d0ccd0",
              "height": 75,
              "marginBottom": 20,
              "width": "100%",
            }
          }
        />
        <View
          style={
            Object {
              "backgroundColor": "#d0ccd0",
              "height": 75,
              "marginBottom": 20,
              "width": "100%",
            }
          }
        />
        <View
          style={
            Object {
              "backgroundColor": "#d0ccd0",
              "height": 75,
              "marginBottom": 20,
              "width": "100%",
            }
          }
        />
        <View
          style={
            Object {
              "backgroundColor": "#d0ccd0",
              "height": 75,
              "marginBottom": 20,
              "width": "100%",
            }
          }
        />
        <View
          style={
            Object {
              "height": 0,
            }
          }
        />
      </View>
    </RCTScrollView>
    <View
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "height": "100%",
          "opacity": 1,
          "width": "20%",
        }
      }
    >
      <Text
        style={
          Object {
            "textAlign": "center",
          }
        }
      >
        
      </Text>
    </View>
  </View>
</View>
`;

exports[`<Menu /> renders correctly on right 1`] = `
<View
  style={
    Object {
      "height": "100%",
      "left": 750,
      "position": "absolute",
      "width": "100%",
      "zIndex": 50,
    }
  }
>
  <View
    style={
      Object {
        "backgroundColor": "rgba(255, 255, 255, 0)",
        "flex": 1,
        "flexDirection": "row",
      }
    }
  >
    <View
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "height": "100%",
          "opacity": 1,
          "width": "20%",
        }
      }
    >
      <Text
        style={
          Object {
            "textAlign": "center",
          }
        }
      >
        
      </Text>
    </View>
    <RCTScrollView
      style={
        Object {
          "backgroundColor": "#bcab79",
          "paddingTop": 0,
          "width": "80%",
        }
      }
    >
      <View>
        <View
          style={
            Object {
              "backgroundColor": "#d0ccd0",
              "height": 75,
              "marginBottom": 20,
              "width": "100%",
            }
          }
        />
        <View
          style={
            Object {
              "backgroundColor": "#d0ccd0",
              "height": 75,
              "marginBottom": 20,
              "width": "100%",
            }
          }
        />
        <View
          style={
            Object {
              "backgroundColor": "#d0ccd0",
              "height": 75,
              "marginBottom": 20,
              "width": "100%",
            }
          }
        />
        <View
          style={
            Object {
              "backgroundColor": "#d0ccd0",
              "height": 75,
              "marginBottom": 20,
              "width": "100%",
            }
          }
        />
        <View
          style={
            Object {
              "backgroundColor": "#d0ccd0",
              "height": 75,
              "marginBottom": 20,
              "width": "100%",
            }
          }
        />
        <View
          style={
            Object {
              "backgroundColor": "#d0ccd0",
              "height": 75,
              "marginBottom": 20,
              "width": "100%",
            }
          }
        />
        <View
          style={
            Object {
              "backgroundColor": "#d0ccd0",
              "height": 75,
              "marginBottom": 20,
              "width": "100%",
            }
          }
        />
        <View
          style={
            Object {
              "backgroundColor": "#d0ccd0",
              "height": 75,
              "marginBottom": 20,
              "width": "100%",
            }
          }
        />
        <View
          style={
            Object {
              "backgroundColor": "#d0ccd0",
              "height": 75,
              "marginBottom": 20,
              "width": "100%",
            }
          }
        />
        <View
          style={
            Object {
              "backgroundColor": "#d0ccd0",
              "height": 75,
              "marginBottom": 20,
              "width": "100%",
            }
          }
        />
        <View
          style={
            Object {
              "backgroundColor": "#d0ccd0",
              "height": 75,
              "marginBottom": 20,
              "width": "100%",
            }
          }
        />
        <View
          style={
            Object {
              "backgroundColor": "#d0ccd0",
              "height": 75,
              "marginBottom": 20,
              "width": "100%",
            }
          }
        />
        <View
          style={
            Object {
              "backgroundColor": "#d0ccd0",
              "height": 75,
              "marginBottom": 20,
              "width": "100%",
            }
          }
        />
        <View
          style={
            Object {
              "backgroundColor": "#d0ccd0",
              "height": 75,
              "marginBottom": 20,
              "width": "100%",
            }
          }
        />
        <View
          style={
            Object {
              "backgroundColor": "#d0ccd0",
              "height": 75,
              "marginBottom": 20,
              "width": "100%",
            }
          }
        />
        <View
          style={
            Object {
              "backgroundColor": "#d0ccd0",
              "height": 75,
              "marginBottom": 20,
              "width": "100%",
            }
          }
        />
        <View
          style={
            Object {
              "backgroundColor": "#d0ccd0",
              "height": 75,
              "marginBottom": 20,
              "width": "100%",
            }
          }
        />
        <View
          style={
            Object {
              "backgroundColor": "#d0ccd0",
              "height": 75,
              "marginBottom": 20,
              "width": "100%",
            }
          }
        />
        <View
          style={
            Object {
              "backgroundColor": "#d0ccd0",
              "height": 75,
              "marginBottom": 20,
              "width": "100%",
            }
          }
        />
        <View
          style={
            Object {
              "backgroundColor": "#d0ccd0",
              "height": 75,
              "marginBottom": 20,
              "width": "100%",
            }
          }
        />
        <View
          style={
            Object {
              "height": 0,
            }
          }
        />
      </View>
    </RCTScrollView>
  </View>
</View>
`;
